src/Microsoft.AspNetCore.Routing/Matchers/CandidateSet.cs

```diff
@@ -2,56 +2,154 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.AspNetCore.Routing.Matchers
 {
-    internal class CandidateSet
+    public sealed class CandidateSet
```

Comment:
Personally, I'd expect a type called CandidateSet to implement ISet.

Reply:
Do you have a suggestion? I have no plans to introduce `ISet` or other high level collection APIs here due to pref cost.

Reply:
As far as I understand it, it's a set (duplicate candidatestates wouldn't make sense (to me)), so the name is better than anything I can think of.